docs/rfcs/022-pageserver-delete-from-s3.md

```diff
@@ -0,0 +1,251 @@
+# Deleting pageserver part of tenants data from s3
+
+Created on 08.03.23
+
+## Motivation
+
+Currently we dont delete pageserver part of the data from s3 when project is deleted. (The same is true for safekeepers, but this outside of the scope of this RFC).
+
+This RFC aims to spin a discussion to come to a robust deletion solution that wont put us in into a corner for features like postponed deletion (when we keep data for user to be able to restore a project if it was deleted by accident)
+
+## Summary
+
+TLDR; There are two options, one based on control plane issuing actual delete requests to s3 and the other one that keeps s3 stuff bound to pageserver. Each one has its pros and cons.
```

Comment:
Threading a new for resulting solution. 

If we don't need to have a non-api, very soft dependency between the projects then lets not have it (s3 layout). Adding one would tie releases and everything, and it will hurt even in short run. Though I guess that could be worked around by having a s3-util supplied by pageserver?

So I'd say we go for pageserver managed. There might be some missing corner cases, but getting the ability to even tombstone timelines then tenants in s3 would free us up of more errors.

Reply:
> Though I guess that could be worked around by having a s3-util supplied by pageserver?

There can be options. As I said [here](https://github.com/neondatabase/neon/pull/3792#discussion_r1135857242):

> There is an option to ask pageserver to convert tenant id into a path on s3 so console can chew the data, but that anyway requires adding s3 related machinery to console (at least a corresponding library)

Path in s3 is a build from prefix specified in pageserver config. Theoretically it can be different for different pageservers. Currently the value is the same everywhere.

***

I'm plus one to start implementing deleted flag in index_part.json. It should solve the immediate problem with attach we have.

In general I like the idea containing this as much as possible inside pageserver. Retry logic will leak to control plane anyway, but it looks like not a big problem. In the long run there were plans to split control plane from console and move control plane to neon repo. This may change the picture a bit, bu still poking around s3 in both binaries may not be a good idea.

Aside from that approach based on control plane can be benefitial in terms of code reuse between pageserver and safekeeper delete impls. It can construct a path (or ask for it from the node) and `ls + delete` everything inside. Currently for safekeepers we dont delete anything from s3 either. And it can be worse than it is for pageserver because there is no gc/PITR, so it is always full history from the beginning of the universe.